File: src/android/com/eclipsesource/firebase/messaging/MessagingOperator.java

package com.eclipsesource.firebase.messaging;

import android.app.Activity;

import com.eclipsesource.tabris.android.AbstractOperator;
import com.eclipsesource.tabris.android.Properties;
import com.eclipsesource.tabris.android.PropertyHandler;
import com.eclipsesource.tabris.android.TabrisContext;

public class MessagingOperator extends AbstractOperator<Messaging> {

  private static final String TYPE = "com.eclipsesource.firebase.Messaging";
  private static final String METHOD_RESET_INSTANCE_ID = "resetInstanceId";

  private final Activity activity;
  private final TabrisContext tabrisContext;
  private final MessagingPropertyHandler propertyHandler;

  public MessagingOperator( Activity activity, TabrisContext tabrisContext ) {
    this.activity = activity;
    this.tabrisContext = tabrisContext;
    propertyHandler = new MessagingPropertyHandler();
  }

  @Override
  public PropertyHandler<Messaging> getPropertyHandler( Messaging messaging ) {
    return propertyHandler;
  }

  @Override
  public String getType() {
    return TYPE;
  }

  @Override
  public Messaging create( String id, Properties properties ) {
    return new Messaging( activity, tabrisContext );
  }

  @Override
  public Object call( Messaging messaging, String method, Properties properties ) {
    if( method.equals( METHOD_RESET_INSTANCE_ID ) ) {
      messaging.resetInstanceId();
    }
    return null;
  }

  @Override
  public void destroy( Messaging messaging ) {
    messaging.unregisterAllListeners();
  }

}

File: src/android/com/eclipsesource/firebase/messaging/MessagingOperator.kt
package com.eclipsesource.firebase.messaging

import android.app.Activity
import com.eclipsesource.tabris.android.AbstractOperator
import com.eclipsesource.tabris.android.Properties
import com.eclipsesource.tabris.android.TabrisContext

private const val TYPE = "com.eclipsesource.firebase.Messaging"
private const val METHOD_RESET_INSTANCE_ID = "resetInstanceId"
private const val METHOD_GET_ALL = "getAllPendingMessages"
private const val METHOD_CLEAR_ALL = "clearAllPendingMessages"

class MessagingOperator(private val activity: Activity, private val tabrisContext: TabrisContext)
  : AbstractOperator<Messaging>() {

  private val propertyHandler by lazy { MessagingPropertyHandler() }

  override fun getType() = TYPE

  override fun getPropertyHandler(messaging: Messaging) = propertyHandler

  override fun create(id: String, properties: Properties) = Messaging(activity, tabrisContext)

  override fun call(messaging: Messaging, method: String, properties: Properties): Any? = when (method) {
    METHOD_RESET_INSTANCE_ID -> messaging.resetInstanceId()
    METHOD_CLEAR_ALL -> messaging.clearAllPendingMessages()
    METHOD_GET_ALL -> messaging.getAllPendingMessages()
    else -> null
  }

  override fun destroy(messaging: Messaging) = messaging.unregisterAllListeners()

}
